meta-mender-core/classes/mender-part-images.bbclass

```diff
@@ -74,14 +74,39 @@ mender_part_image() {
     wks="${WORKDIR}/mender-$suffix.wks"
     rm -f "$wks"
     if [ -n "${IMAGE_BOOTLOADER_FILE}" ]; then
+
+        spl_align_kb=0
+        if [ -n "${SPL_BINARY}" ]; then
+            if [ -z "${MENDER_SPL_OFFSET_KB}" ] ; then
+                bberror "MENDER_SPL_OFFSET unspecified.  Please consult your board" \
+                        "to determine the correct value."
+            fi
+
+            # Copy the files to embed in the WIC image into ${WORKDIR} for exclusive access
+            install -m 0644 "${DEPLOY_DIR_IMAGE}/${SPL_BINARY}" "${WORKDIR}/"
+
+            spl_align_kb=${MENDER_SPL_OFFSET_KB}
+            spl_size=$(stat -c '%s' "${WORKDIR}/${SPL_BINARY}")
+            spl_end=$(expr $spl_align_kb \* 1024 + $spl_size)
+            if [ $spl_end -gt $(expr ${IMAGE_BOOTLOADER_OFFSET} \* 1024) ]; then
+                bberror "Size of SPL specified in SPL_BINARY is too large." \
+                        "It will be overwritten by the IMAGE_BOOTLOADER_FILE." \
+                        "Please raise it manually."
+            fi
+            cat >> "$wks" <<EOF
+# embed SPL
+part --source rawcopy --sourceparams="file=${WORKDIR}/${SPL_BINARY}" --ondisk mmcblk0 --align $spl_align_kb --no-table
+EOF
```

Comment:
Why is this needed in addition to `IMAGE_BOOTLOADER_FILE`? Not sure I understand the context here.

Reply:
Some U-Boot systems require a multi-stage setup.  Within U-Boot there is SPL (secondary program loader) and TPL (tertiary program loader).  I don't see any support in Yocto for TPL but there is definitely support for SPL on certain platforms.  Beaglebone actually uses it via a file in the VFAT partition (filename MLO).  In the case of the udooneo platform, there is a small SPL binary located in the raw disk image that then loads the primary U-Boot binary.  See https://github.com/graugans/meta-udoo/blob/morty/scripts/lib/wic/canned-wks/udoo-sdcard.wks for context.

Reply:
This has actually been discussed on the mailing list, @kacf @drewmoseley.  

I have been in the situation where I need to support two files for boot-loader (SPL and u-boot.bin) as well. My workaround has been to simply `dd` the two seperate files in to one in a "post_install" step in the U-boot recipe and provide combined output file to `IMAGE_BOOTLOADER_FILE`.

What we also discussed on that particular thread was to add support for passing multiple files to `IMAGE_BOOTLOADER_FILE` instead of adding a second argument.

I believe the suggestion was something like this (do not recall the exact syntax but I think the idea was the same):

```
IMAGE_BOOTLOADER_FILE = "SPL,1; u-boot.bin,69;"
```

And that is filename together with an offset. That way we can support a TPL as well or other binaries that need to be put at an offset on the storage medium.

Reply:
Found the thread: https://groups.google.com/a/lists.mender.io/forum/#!searchin/mender/mirza$20krak/mender/hyHk9a6C4K8/A23cC8qLBAAJ

Reply:
@mirzak: Oh yes, I had completely forgotten about that. Nice catch! I would definitely prefer this approach because it is much more generic.

Tweaking it ever so slightly, I would propose the syntax:

```
IMAGE_BOOTLOADER_FILE = "SPL;1 u-boot.bin;69"
```

for two reasons:

* Similarity in to [`IMAGE_BOOT_FILES`](https://www.yoctoproject.org/docs/latest/mega-manual/mega-manual.html#var-IMAGE_BOOT_FILES), with `src;location` format.
* Using only spaces for separating entries to ease shell parsing.

The great thing about this approach is that we can support as many special files as people want.

Reply:
I like this suggestion.  Not sure if I'll have time to implement it anytime soon though.  I created https://tracker.mender.io/browse/MEN-1883 to track it.